torch/quantization/_numeric_suite_fx.py

```diff
@@ -137,3 +137,204 @@ def compare_weights_fx(float_dict, quantized_dict):
+
+
+def get_logger_dict(mod, prefix=""):
+    r"""Traverse the modules and save all logger stats into target dict.
+    This is mainly used for quantization accuracy debug.
+
+    Type of loggers supported:
+        ShadowLogger: used to log the outputs of the quantized module and its
+            matching float shadow module,
+        OutputLogger: used to log the outputs of the modules
+
+    Args:
+        mod: module we want to save all logger stats
+        prefix: prefix for the current module
+
+    Return:
+        target_dict: the dictionary used to save all logger stats
+    """
+    torch._C._log_api_usage_once("quantization_api._numeric_suite.get_logger_dict")
+
+    target_dict: Dict[str, Dict] = {}
+    _get_logger_dict_helper(mod, target_dict, prefix)
+    return target_dict
+
+
+class Logger(nn.Module):
+    r"""Base class for stats logging"""
+
+    def __init__(self):
+        super(Logger, self).__init__()
+        self.stats = {}
+
+    def forward(self, x):
+        pass
+
+
+class ShadowLogger(Logger):
+    r"""Class used in Shadow module to record the outputs of the original and
+    shadow modules.
+    """
+
+    def __init__(self):
+        super(ShadowLogger, self).__init__()
+        self.stats["float"] = []
+        self.stats["quantized"] = []
+
+    def forward(self, x, y):
+        if len(x) > 1:
+            x = x[0]
+        if len(y) > 1:
+            y = y[0]
+        self.stats["quantized"].append(x.detach())
+        self.stats["float"].append(y.detach())
+
+
+def _convert_tuple_to_list(t):
+    return list(_convert_tuple_to_list(x) for x in t) if type(t) is tuple else t
+
+
+def _dequantize_tensor_list(t):
+    return (
+        list(_dequantize_tensor_list(x) for x in t)
+        if type(t) is list
+        else t.dequantize()
+        if t.is_quantized
+        else t
+    )
+
+
+class Shadow(nn.Module):
+    r"""Shadow module attaches the float module to its matching quantized module
+    as the shadow. Then it uses Logger module to process the outputs of both
+    modules.
+
+    Args:
+        q_module: module quantized from float_module that we want to shadow
+        float_module: float module used to shadow q_module
+        Logger: type of logger used to process the outputs of q_module and
+            float_module. ShadowLogger or custom loggers can be used.
+    """
+
+    def __init__(self, q_module, float_module, Logger):
+        super(Shadow, self).__init__()
+        self.orig_module = q_module
+        self.shadow_module = float_module
+        self.dequant = nnq.DeQuantize()
+        self.logger = Logger()
+
+    def forward(self, *x):
+        xl = _convert_tuple_to_list(x)
+        output = self.orig_module(*xl)
+        xl_float = _dequantize_tensor_list(xl)
+        shadow_output = self.shadow_module(*xl_float)
+        self.logger(output, shadow_output)
+        return output
+
+
+def prepare_model_with_stubs_fx(float_module, q_module, module_swap_list, Logger):
```

Comment:
this is pretty similar to eager mode, can you just call eager mode api on a fx graphmode quantized model for this?

Reply:
Yes, they are the same in eager mode.  My inclination is to keep fx version of prepare_model_xxx api and compare_model_xxx separate from eager mode.  Because for another fx api: compare_model_output_fx , prepare_model_output is different from eager mode and we have the API in fx.  So to make it consistent, I was thinking to  have it reside in fx.

Reply:
Since prepare_model_with_stubs_fx() will be a top level API users will use, maybe good to keep "_fx" suffix to be consistent with other fx numeric suite APIs? Also it would be more flexible if we want to make fx specific changes to this function in the future.

Reply:
oh I mean you can call eager mode prepare_model_with_stubs here without copying the code. I think the actual fx implementation will not be able to reuse this code any ways

Reply:
Oh I see. We can make it as a wrapper for prepare_model_with_stubs for now, and make specific changes to this function in the future if requires so. How does that sound?

Reply:
yes, that sounds good, I think we'll need to re-implement in the future when we want to enable some features that can only be supported by fx

Reply:
Makes sense, we could still reuse the eager code for modules and add additional fx specific code in the body of this function